63 replace our mock data with the core data sent from customer

michaelbrusegard opened the pull request:
WIP, don't merge yet. I created the PR early for discussion purposes

michaelbrusegard committed:
add first iteration of EmojiRepository

michaelbrusegard committed:
work on migrating to viewmodels

michaelbrusegard committed:
feat: finish EmojiRepository

michaelbrusegard committed:
add news repository and mockdata

michaelbrusegard committed:
add news repository to application

michaelbrusegard committed:
update Activity and Application to not have emoji since it isnt related to the feature

michaelbrusegard committed:
fix rename application & activity

michaelbrusegard committed:
add ViewModels for categories and catalogue

michaelbrusegard committed:
worked on implementing the new EmojiCatalogueViewModel

michaelbrusegard committed:
add categories

michaelbrusegard committed:
Merge remote-tracking branch 'origin/dev' into 63-replace-our-mock-data-with-the-core-data-sent-from-customer
* origin/dev:
  fix: :pencil2: changed the name of test function to something simpler in CategoryScrollCarouselTest.kt
  test: ✅ Made UI test for CategoryScrollCarousel.kt
  set up admob

michaelbrusegard committed:
add method to get popular emojis

michaelbrusegard committed:
rename coroutine to better reflect that it returns a list of emojis and not a map

michaelbrusegard committed:
add events mockdata and EventRepository

michaelbrusegard committed:
add EventsRepository to Application

OleAndreasER committed:
view model progress

OleAndreasER committed:
Merge branch '63-replace-our-mock-data-with-the-core-data-sent-from-customer' of https://github.com/ITP2-G15/Open-Emojiplatform into 63-replace-our-mock-data-with-the-core-data-sent-from-customer

OleAndreasER committed:
search screen and emoji details work somewhat

OleAndreasER committed:
most popular

OleAndreasER committed:
fromCategory -> ofCategory, and is now a pair not map

OleAndreasER committed:
state-managing component versions get more specific names

OleAndreasER committed:
fix emoji repository delay simulation issues

OleAndreasER committed:
live data -> state flow, faster and nicer looking

michaelbrusegard commented on app/src/main/kotlin/com/platform/openemoji/emoji/EmojiRepository.kt:
```diff
@@ -0,0 +1,84 @@
+        return emojisOfCategory
+    }
+
+    suspend fun getOverviewEmojis(limitPerCategory: Int): Map<String, List<Emoji>> {
```
@OleAndreasER I see that you removed a bunch of the caching, did it create issues

OleAndreasER replied:
I did. It created some problems with simulating delay. I'm caching a different way now. I also think it's a bad idea to try to anticipate their API to set up caching. They might use pages or limit and offset or something different.

michaelbrusegard replied:
okay, but the way it is set up now it will load all new emojis when a category is opened, while before it didnt reload the ones provided from the overview

OleAndreasER replied:
Yes, I will explain later after Smash bros.

OleAndreasER replied:
You would need to reload them anyway unless you use limit + offset or pages. That's why we shouldn't think about it too much yet. There might a reason for them to prefer a certain system. Also, I remember there was an inbuilt cache system in the GraphQL client library I used before, and they might want to use something similar. The cache should be more structured than a title -> emoji map for quicker access. All in all there are uncertainties that make it not worth to anticipate these things too much.

michaelbrusegard commented:
@OleAndreasER Do you want me to take a look at any of the viewmodels to help on this now? Or update any of the repositories if it is needed

OleAndreasER commented:
@michaelbrusegard I'm working on something now. I can tell you when I have something that you can evaluate. Don't commit anything.

OleAndreasER committed:
merge category into catalogue viewmodel for better coordination

OleAndreasER commented:
@michaelbrusegard you can look now. Just ask if there's any decisions you are unsure about. I'll continue to work on stuff.

OleAndreasER committed:
fix some tests

OleAndreasER committed:
Merge branch 'dev' into 63-replace-our-mock-data-with-the-core-data-sent-from-customer

michaelbrusegard commented on app/src/main/res/values/strings.xml:
```diff
@@ -13,7 +13,6 @@ easily add support for multiple languages if needed -->
     <string name="search_for_emoji">Search for emoji.</string>
     <string name="search_label">Search for emoji</string>
     <string name="search_result">Search results for:</string>
-    <string name="overview">Overview</string>
```
@OleAndreasER Do we not need the overviewstring anymore?

OleAndreasER replied:
Since the string has meaning in the code beyond the composables, I think it makes sense for it to be a constant. It can't be accessed easily from R.string outside of composables.

michaelbrusegard replied:
I see, if it is visible in the UI we should have it stored in the xml file

OleAndreasER committed:
news viewmodel

OleAndreasER committed:
event viewmodel

OleAndreasER committed:
improve search speed + bebounce

OleAndreasER committed:
improve loading

OleAndreasER committed:
fix multiple loadings of catalogue bug

michaelbrusegard commented:
@OleAndreasER for the search we can add a call delay so it doesnt call the repository for every input change. It calls it once when it changes and the waits for lets say 200-300ms before taking the input and calling again. I think this should be an easy to use built in functionality we can add

michaelbrusegard committed:
add missing name properties from emojis

michaelbrusegard committed:
fix 'ImageVector' is deprecated for icons

michaelbrusegard committed:
parse description content as html + responsive design for emoji sequences with column layout

michaelbrusegard commented:
<img width="375" alt="Screenshot 2024-03-30 at 01 54 52" src="https://github.com/ITP2-G15/OpenEmojiPlatform/assets/56915010/fb313bad-cb6b-4f62-8e91-fa4cd899f430">
<img width="369" alt="Screenshot 2024-03-30 at 01 56 20" src="https://github.com/ITP2-G15/OpenEmojiPlatform/assets/56915010/169387f3-b43e-4deb-9d03-4c0e284cf764">

HTML gets parsed so the description is readable. also changed the padding and moved to a column layout to make it work with sequences

michaelbrusegard commented:
@OleAndreasER I manually added the missing name fields to the emojis.json, but there are still fields missing like description for the phoenix. The data they gave us is incomplete, but as long as all emojis have a name the app will work correctly with navigation

michaelbrusegard committed:
comment overview string in EmojiRepository

michaelbrusegard commented:
After reading the code i understand better why you moved "overview" out of strings.xml. It is better to be handled in the backend than in the UI if a seperate overview view is present. We do it now in the repository but you get what i mean

michaelbrusegard committed:
Merge remote-tracking branch 'origin/dev' into 63-replace-our-mock-data-with-the-core-data-sent-from-customer
* origin/dev:
  added modifier and testtags so that tests passes
  reformated using ktlintformat
  replaced backbutton and showmore with components for it Closes #83

michaelbrusegard commented:
So I think this is almost ready, what is missing is tests and loading UI? Anything else? We can move those into new issues to get it merged. wdyt?

michaelbrusegard committed:
sort overview emojis by popularity

michaelbrusegard commented:
> @OleAndreasER for the search we can add a call delay so it doesnt call the repository for every input change. It calls it once when it changes and the waits for lets say 200-300ms before taking the input and calling again. I think this should be an easy to use built in functionality we can add

Forget this, you had already added it to the viewmodel :))

michaelbrusegard committed:
add EmojiRepositoryTest

michaelbrusegard committed:
add EventsRepositoryTest

michaelbrusegard committed:
add NewsRepositoryTest

michaelbrusegard commented:
@OleAndreasER I added tests for the repositories. Can you add for the viewmodels? Also do we need to maybe change the AndroidTest stuff, dont know if it works

OleAndreasER commented:
@michaelbrusegard Nice work! I'll work on testing today. Loading ui should be another issue. Let's say this issue is finished after tests. There are minor improvements here and there that can be future issues.

OleAndreasER committed:
repository interfaces

michaelbrusegard commented:
why do we need interfaces for the repositories

OleAndreasER commented:
@michaelbrusegard It's not necessary. We could use our repository implementations set up with your (e.g.) testEmojis parameter. But I think how I'm gonna do will be nicer for testing of viewmodels and components. We'll see.

michaelbrusegard commented:
> @michaelbrusegard It's not necessary. We could use our repository implementations set up with your (e.g.) testEmojis parameter. But I think how I'm gonna do will be nicer for testing of viewmodels and components. We'll see. 

ok

OleAndreasER committed:
fix navigation tests

OleAndreasER commented:
@michaelbrusegard I think this is a very flexible approach. Look at navigation tests.

OleAndreasER committed:
all tests pass

michaelbrusegard commented:
do we have to also add viewmodel tests

OleAndreasER commented:
@michaelbrusegard Actually we probably don't. It might be better to test them indirectly by testing the screens. I'm trying to test the view models now, but it feels a little silly.

OleAndreasER commented:
I need to fix search test.

michaelbrusegard commented:
> I need to fix search test.

ok, me or you cant review. we should get someone else

OleAndreasER committed:
replace search field test with a simpler one

We should test the search screen later

OleAndreasER committed:
remove test that wasn't supposed to be there

Andreasli02 commented on app/src/main/kotlin/com/platform/openemoji/news/NewsViewModel.kt:
```diff
@@ -0,0 +1,23 @@
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+
+const val NEWS_IN_NEWS_LIST_SCREEN = 15
+
```
MAX_NEWS_FETCHED ?

OleAndreasER replied:
The view model doesn't know about fetching. It knows about the UI, the list screen, sort of, since view models are made for the composables that use them. This constant is temporary, probably. Should be state in the view model, and it should increase as the user loads more news.

Steffenriv replied:
I agree the name can be a bit confusing maybe change to NEWS_CARDS or something else entirely?! I had the same idea as you Ole which i wrote earlier on the Events view model, by having a cache and load as user scrolls. This is a change we need to do later, so you guys do not need to think about it now!

Steffenriv replied:
Also do not think this constant is needed as there will very few newscards added to newsListScreen. Although i can see the reason why you added it.

Andreasli02 commented on app/src/main/kotlin/com/platform/openemoji/events/EventViewModel.kt:
```diff
@@ -0,0 +1,23 @@
+import kotlinx.coroutines.launch
+
+const val NEWS_IN_EVENT_LIST_SCREEN = 15
+
```
MAX_EVENTS_FETCHED ?

Andreasli02 reviewed (commented):
Cant do a very good review since I dont know very much about viewmodels and repositories. Look great though. 

Additional suggestion:
Implement a minimum 2-character length requirement for search queries?
Think error handling is normal and can be added?

OleAndreasER commented:
Gonna add loading icon spinner thing and text for no search results.

OleAndreasER committed:
loading indicator and no search result text

OleAndreasER committed:
vertical scroll on search results

Steffenriv commented:
> Gonna add loading icon spinner thing and text for no search results.

I suppose this is added to relate to skeleton view issue as we spoke of last week?

Steffenriv commented on app/src/androidTest/kotlin/com/platform/openemoji/navigation/NavigationTest.kt:
```diff
@@ -68,24 +85,4 @@ class NavigationTest {
             .performClick()
         composeTestRule.onNodeWithText("Search results for: h").assertExists()
     }
```
Have you guys merged (or rebased) then deleted lines of code that you have not written? Multiple places have added deleted and added full files, as well as deleted code that is already in dev. Just wondering, this will not cause any issues. NO CHANGE NEEDED!

michaelbrusegard replied:
I did a bunch of merge conflicts. Had to rewrite a bunch of stuff to make it work with MVVM architecture. For example EmojiCatalogueUI -> EmojiCatalogue now because it makes sense. And the old EmojiCatalogue is basicly the viewmodel

Steffenriv replied:
Alright, it seems to be fixed now so that's good! Just wanted to know what had happened, thought that might be the case.

Steffenriv commented on app/src/main/kotlin/com/platform/openemoji/Application.kt:
```diff
@@ -0,0 +1,27 @@
+import com.platform.openemoji.news.NewsMockDataRepository
+import com.platform.openemoji.news.NewsRepository
+
+interface RepositoryStore {
```
As i do not think everybody understands repositories and viewmodels. I think some kind of documentation as to why this is needed and added would be nice to have (although they could read up on it aswell, maybe add some useful links).

michaelbrusegard replied:
Repositories are for fetching the data and caching it if needed. How they get the data doesnt matter for the viewmodel. The viewmodel holds the data. Composables "subscribes" to the data within a viewmodel. We call methods within the viewmodel if we need to update the data, which will be then updated across all composables which are "subscribed". The reason for a viewmodel is so that when updates happen in the UI we wont have to refetch the data because it is in the viewmodel

Steffenriv replied:
... I understand just fine. Mostly referring to when others might want to use them, no need to be very detailed, just a quick explaination in code (perhaps a link if that is too hard). Just so everybody can understand it properly! If you do not like this, you could always add this in the readme.

michaelbrusegard replied:
We can add to docs link in Notion maybe? There is a website on android.dev that explains a lot about MVVM

Steffenriv replied:
We need to write about it in the report anyways, so i think some kind of short documentation in someway could turn out to be nice especially since we are delivering the code!

OleAndreasER replied:
We should explain the general idea of view models and repositories at Zedge to everyone. In general, I don't think comments in the code should explain how libraries or the programming language works (as long they are being used in standard ways). There are probably still some more good comments that can be added. I will go over the code again.

Steffenriv replied:
Sounds like a plan! You can ignore documentation then.

Steffenriv commented on app/src/main/kotlin/com/platform/openemoji/emoji/EmojiIcon.kt:
```diff
@@ -22,36 +22,22 @@ import androidx.compose.ui.text.AnnotatedString
 import com.platform.openemoji.R
 import com.platform.openemoji.navigation.Screen
 
-// This will only be used for design variants in the emoji icon view if we have time to implement that
```
Why was this deleted?

michaelbrusegard replied:
Ask @OleAndreasER I think he did it. it isnt being used right now, we need emoji design variant data from customer to have a use for it. We can request it

Steffenriv replied:
I think we should request it, could turn out useful when implementing technical details! However i can understand deleting code that isn't used.

michaelbrusegard replied:
i agree should be kept and we request from customer

OleAndreasER replied:
Can't remember if I deleted it or not. But since it's not used, I don't think it should be there.

OleAndreasER replied:
Oh emojis don't have image urls anymore.

Steffenriv replied:
We can add it back later if necessary! So we don't need to think further about it, we can fix that later if need be.

Steffenriv commented on app/src/main/kotlin/com/platform/openemoji/emoji/EmojiRepository.kt:
```diff
@@ -0,0 +1,130 @@
+interface EmojiRepository {
+    suspend fun getCategories(): List<String> = emptyList()
+
+    suspend fun getEmojisOfCategory(category: String): Map<String, List<Emoji>> =
```
Still think getEmojisByCategory is more suitable.

michaelbrusegard replied:
I am not sure but i think it is named to not confuse with the similarly named viewmodel method.

Steffenriv replied:
No such named method in viewmodel!

michaelbrusegard replied:
ask @OleAndreasER He redid a bunch of stuff in the viewmodel

OleAndreasER replied:
This does not return emojis by category; it returns emojis only of the single category. This method needs a clarifying comment.

Steffenriv replied:
Uhm, they literally mean the same thing. There is no difference to the naming, i just feel like by is more regularly used therefore also better to use.

OleAndreasER replied:
Using "byCategory" implies that there are more categories, or at least that there might be. "ofCategory" tells you that there is only one.

Steffenriv replied:
Disagree, but it isn't that important you left a comment about it so i'll leave it at that.

Steffenriv commented on app/src/main/kotlin/com/platform/openemoji/emoji/EmojiRepository.kt:
```diff
@@ -0,0 +1,130 @@
+
+    suspend fun getPopularEmojis(limit: Int = Int.MAX_VALUE): List<Emoji> = emptyList()
+
+    suspend fun getEmoji(name: String): Emoji? = null
```
getEmojiByName? (since all getters have specifications I don't think shortening this is suitable)

OleAndreasER replied:
Do you mean the other getters in this repository? "getPopularEmojis", "getEmojisOfCategory"? "OfCategory" in that last one is there to explain why it returns a Map instead of a List, which you would expect if it was named "getEmojis". Naming the method "getEmojiByName" is more like naming "searchEmojis(query: String)" "searchEmojisByQuery(query: String)". If you insist, I'll change it. I'm not too sure.

Steffenriv replied:
Honestly, i feel like searchEmojis is also a bad name for it, but did not have a good replacement name so did not comment on it. However, i feel if we use names such as getPopularEmojis getEmojisOf/ByCategory, then we should not name them without specifying either. Not that it isn't understandable as they are, just so that we follow the same format!

OleAndreasER replied:
The difference is that getPopularEmojis and getEmojisOfCategory describe the return value. "getEmoji(name)" and "getEmojiByName(name)" say the same thing about what is returned; the latter just repeats what is needed to get the emoji, the emoji name, which is already described by the parameter.

Steffenriv replied:
I guess it is fine then, hoped to get them changed, but it really isn't that important i guess.

Steffenriv commented on app/src/main/kotlin/com/platform/openemoji/emoji/EmojiRepository.kt:
```diff
@@ -0,0 +1,130 @@
+                withContext(Dispatchers.IO) {
+                    val inputStream = context.assets.open("emojis.json")
+                    val size = inputStream.available()
+                    val buffer = ByteArray(size)
```
I do think some more documentation would be nice here as well! Since this can be confusing to get grasp of at first read through.

michaelbrusegard replied:
the stuff related to context is reading the data from file and is not needed anymore when the customer uses their own api

Steffenriv commented on app/src/main/kotlin/com/platform/openemoji/events/EventViewModel.kt:
```diff
@@ -0,0 +1,23 @@
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+
+const val NEWS_IN_EVENT_LIST_SCREEN = 15
```
This name does not make sense, remove news. Instead use at least EVENTS or perhaps EVENT_CARDS! Also not sure this is needed! Unlike emojis, we do not get enough events that this should be an issue. What we can do instead, is create a cache storing the first X upcoming events (where this could be needed) and then load more events if the user goes further down. How we store events (yearly or further) is yet to be decided and is something i think the client can decide.

Steffenriv replied:
Change name at least! No need to adjust to using cache just yet.

michaelbrusegard replied:
ask @OleAndreasER

Steffenriv commented on app/src/main/kotlin/com/platform/openemoji/events/EventsRepository.kt:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    override suspend fun getEvents(limit: Int): List<Event> {
+        val allNews = loadMockdata()
```
again is allNews does not make sense! Has there been confusion between events and news, or is same code used perhaps with forgotten changes?

michaelbrusegard replied:
my bad, must have forgot to rename when using similar code in the news and events repositories. Will fix when I get home

Steffenriv commented on app/src/main/kotlin/com/platform/openemoji/search/EmojiSearchViewModel.kt:
```diff
@@ -0,0 +1,39 @@
+            .debounce(250)
+            .mapLatest {
+                if (it.isNotEmpty()) {
+                    searchResultsAreLoading.value = true
```
I would probably have added documentations as to detail what this searchQuery is doing, but you can decide on that (i may be a bit strict with documentations hehe)

michaelbrusegard replied:
we didnt have documentation on it before, what exactly should be documented do you think

Steffenriv replied:
Just writing something about return values and use of loading check. No need for something long, it kinda is obvious just thought it was a bit confusing when looking at the code the first time (especially when using (it.) preferably i use -> arrows to rename it so it explains what is actually checked.  ( this case string, so not that hard to understand, but since it is used quite a lot, it can become confusing to understand which values it contains.)  At least this is just some yapping thoughts from me.

OleAndreasER replied:
It needs a comment, I agree.

Steffenriv replied:
nice update, much more clear now

Steffenriv commented on app/src/main/kotlin/com/platform/openemoji/emoji/EmojiRepository.kt:
```diff
@@ -0,0 +1,130 @@
+
+// The overview category is a special category that contains a limited number of emojis from each category
+// This can also be retrieved from the API or changed to a different view
+const val OVERVIEW = "Overview"
```
why not use stringResource?! (Because it isn't used directly in Text() composable?

michaelbrusegard replied:
I thought the same. but since it is in the repository it will look like "Overview" is part of the API in the composables which I think makes sense. If we move it to the Composables or ViewModel i think it should be a stringresource

Steffenriv replied:
I do still think that this should use stringResource() as category name should be able to be changed depending on languages later, although we do not yet have this implementation.

michaelbrusegard replied:
I am 50/50 on this. Since their API can have the option to change language or not and they may put overview as part of their API. But they also may not which makes stringresource better

Steffenriv replied:
Having it in stringResource would either way be a better solution then would it not? Since if they do add it to their api, there should be no need for us to declare it as is done now.

OleAndreasER replied:
I changed my mind on this. It should be a string resource. The reason not to use it was that string resource is only accessible in composables. But I will make it so the repository doesn't need to know about the overview string.

Steffenriv reviewed (changes requested):
Well done guys! Code and testing seems to be done in a good way so do not have any direct comments on that. However i do think there could be some more documentation a few places, alongside some name changes (especially replace news as a part of name in variables in event files).

michaelbrusegard committed:
fix news to events typo

michaelbrusegard committed:
Merge branch '63-replace-our-mock-data-with-the-core-data-sent-from-customer' of github.com:ITP2-G15/OpenEmojiPlatform into 63-replace-our-mock-data-with-the-core-data-sent-from-customer

* '63-replace-our-mock-data-with-the-core-data-sent-from-customer' of github.com:ITP2-G15/OpenEmojiPlatform:
  vertical scroll on search results
  loading indicator and no search result text
  remove test that wasn't supposed to be there
  replace search field test with a simpler one

OleAndreasER committed:
commenting and overview to string resources

OleAndreasER committed:
Merge branch '63-replace-our-mock-data-with-the-core-data-sent-from-customer' of https://github.com/ITP2-G15/Open-Emojiplatform into 63-replace-our-mock-data-with-the-core-data-sent-from-customer

Steffenriv commented on app/src/main/kotlin/com/platform/openemoji/emoji/EmojiRepository.kt:
```diff
@@ -37,6 +37,7 @@ class EmojiMockDataRepository(
 ) : EmojiRepository {
     private val catalogueCache: ConcurrentMap<String, Map<String, List<Emoji>>> =
         ConcurrentHashMap()
+    private var overviewEmojisCache: Map<String, List<Emoji>>? = null
```
That variable name feels very strange, any way to change it. Perhaps as before cachedOverviewEmojis.

OleAndreasER committed:
fix constant name
